### PR TITLE
Dynamic exclusivity enforcement for non-escaping closures.

### PR DESCRIPTION
**Pull request by atrick**

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test.

**Review by devincoughlin — COMMENTED** [review]
This looks good to me, although the commented-out disabled verification in tests is a bit weird. Could we pull those out into separate files or use #if?

File: test/SILOptimizer/access_enforcement_noescape.swift [hunk 1]
```diff
@@ -372,15 +339,16 @@ func readBoxWriteInout() {
...
 
 // Error: inout cannot be captured.
-// func inoutReadBoxWriteInout(x: inout Int) {
-//   let c = { _ = x }
-//   doOneInout(c, &x)
-// }
+// Verification disabled because is suppresses other errors.
+//func inoutReadBoxWriteInout(x: inout Int) {
+//  disabled-error@+1{{escaping closures can only capture inout parameters explicitly by value}}
+//  let c = { _ = x }
```
Comment (atrick) [inline test/SILOptimizer/access_enforcement_noescape.swift]: Right. Those commented out tests aren't really specific to this PR. I just happened to notice  that they broke '-verify' if I try to enable them (not sure why). I'll try moving them to a different file and getting them enabled--follow up PR later today.

Comment (atrick) [inline test/SILOptimizer/access_enforcement_noescape.swift]: To answer your first question. The idea behind this file is to cover the entire space involving closure reads/writes along with `inout` arguments on both the caller and callee side. The tests are ordered such that it's easy to demonstrate full coverage. Of course, some of these cases are static errors. I tried mixing -verify and FileCheck patterns to handle both swift 3 and 4. Unfortunately, some compile-time errors prevent later diagnostics from running and reporting the other errors!

**Pull request merged**